_posts/2025-04-22-test-classloading-rewrite.adoc

```diff
@@ -0,0 +1,63 @@
+We're keen to get feedback from the community so that we can get fixing.
+
+== Why?
+
+In previous versions, Quarkus tests were invoked using the default JUnit classloader, and then executed in a different, Quarkus-aware, classloader.
+
+This mostly worked very well, and meant that `QuarkusTest` tests mostly behaved as if they were part of the same application as the code under test.
+The Quarkus test framework could start and stop Quarkus instances at the right point in the test lifecycle, inject CDI dependencies, and do other useful Quarkus bytecode manipulation.
+However, some use cases didn't work. Tests using advanced JUnit features like `@TestTemplate` and `@ParameterizedTest` sometimes found that the same test code might appear to run in several classloaders in a single test, or that injected dependencies weren't always available.
+
+While Quarkus extensions can do all sorts of marvellous bytecode manipulation to improve the developer experience, they cannot manipulate test classes with the same freedom that they do normal application classes.
+
+Over time, test-related defects were building up that couldn't be changed without a fundamental rewrite of test classloading. The Quarkus test code itself was also growing ever-more complex as it tried to work around various JUnit edge cases. Moving test instances from one classloader to another involved serializing and deserialization, which is harder to implement on newer JVM versions with tighter class security. For example, Quarkus used to use XStream as the serialization provider, but XStream no longer works with Java 17 and higher, because of reflection restrictions in the newer JVMs.
+
+What if, instead, Quarkus tests were simply run in the same classloader used to to load them?
+
+== What you need to do
+
+From Quarkus 3.22 onwards, this is exactly how `@QuarkusTest` classloading works.
+What do your tests need to change in order to work with the new architecture?
+Hopefully, nothing.
+One of the goals of this change was that the rewrite didn't touch any tests in our test suite, to make sure they'd all continue working without updates.
+In practice, there have been a few hiccups and we've discovered some edge cases in the broader ecosystem.
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause port conflicts and cross-talk on configuration values. We're hoping to have a https://github.com/quarkusio/quarkus/issues/45785[fix] for this in the next release. As a workaround, splitting conflicting tests into separate projects should fix symptoms.
```

Comment:
discovery / augmentation - we might need a link to a reference guide. I actually never heard about the discovery phase.

Reply:
We may have to explain augmentation somewhere (in an admonition)

Reply:
That's JUnit discovery (oops!). It's a problem with working with JUnit that loads of concepts get overloaded - there's JUnit extensions and Quarkus extensions, JUnit phases and Quarkus phases ... which kind of phase was clear in my head, but no one else would have any way of knowing, so I'll clarify.

I'll also add some links for augmentation.